### PR TITLE
(BKR-658) Fix BEAKER_provision=no when using Docker

### PR DESCRIPTION
**Pull request by aaronbbrown**
Per BKR-658, the Docker provisioner does not behave correctly when `BEAKER_provision=no` is used from beaker-rspec.  This corrects that behavior so the container is reused.  It defers to ::Docker::Container.create throwing raising an exception if `BEAKER_provision=yes` and the container currently exists, rather than blindly reusing the container (the previous behavior). 

Also added is a best effort at sshd_config repair for root logins.  Puppet users will often disable root logins, which breaks Beaker, so this attempts to revert such changes via Docker::Container#exec.

### CODE REVIEW

**Comment by kevpl** [comment]
Hi @aaronbbrown!  Thanks for these commits, we look forward to getting them into the project.

To match up with our CONTRIBUTOR standards, please prefix the git commit messages with the ticket prefix `(BKR-658)`.

@electrical is the community maintainer of the docker hypervisor.  He'll need to take a look at this, and update with any comments or concerns he has before we'll move forward with this.

Once we've gotten past that stage, we'll run through automation to make sure it doesn't break any existing testing. Then we'll merge this into the mainline, to be released on the next train detailed in our [Current Status doc](https://github.com/puppetlabs/beaker/wiki/Current-Status).

Let me know if you have any questions about any of this.

**Comment by aaronbbrown** [comment]
Thanks!  I have squashed my commits and prefixed them per your recommendation.  I look forward to having it reviewed.

File: spec/beaker/hypervisor/docker_spec.rb [hunk 1]
```diff
@@ -305,6 +294,41 @@ module Beaker
         expect( hosts[0]['docker_image'] ).to be === image
         expect( hosts[0]['docker_container'] ).to be === container
       end
+
+      context 'provision=false' do
+        before do
```
Comment (kevpl) [inline spec/beaker/hypervisor/docker_spec.rb]: can this block be deleted?

**Comment by kevpl** [comment]
sorry to nitpick, but we don't usually merge PRs that have merge commits in their own logs.  We usually rebase to get our code up-to-date if needed.

Other than that (and my comment on the code), things look alright to me.  @electrical ?

**Comment by aaronbbrown** [comment]
@kevpl I fixed the commit history and removed the spurious `before` block.

**Comment by kevpl** [comment]
cool!  Automation time!

**Comment by kevpl** [comment]
@puppetlabs-jenkins test this please

**Comment by electrical** [comment]
Sorry, missed the ping of this one. Having a look at it now :-)

File: lib/beaker/hypervisor/docker.rb [hunk 2]
```diff
@@ -68,33 +68,31 @@ def provision
...
+            container = ::Docker::Container.create(container_opts)
           end
         end
 
-        # If the specified container exists, then use it rather creating a new one
-        if container.nil?
-          @logger.debug("Creating container from image #{image_name}")
-          container = ::Docker::Container.create(container_opts)
-        end
+        raise RuntimeError, "Unable to create container because provision=false" if container.nil?
```
Comment (electrical) [inline lib/beaker/hypervisor/docker.rb]: error message could be a bit clearer. Mention something that the container couldn't be found ?
Perhaps fail when you do the find part?

Comment (aaronbbrown) [inline lib/beaker/hypervisor/docker.rb]: How about

> Cannot continue because no existing container could be found and provisioning is disabled.

Comment (aaronbbrown) [inline lib/beaker/hypervisor/docker.rb]: @electrical Can I get feedback on the wording above?

Comment (electrical) [inline lib/beaker/hypervisor/docker.rb]: Sorry, missed the ping. error message sounds good. thanks!

**Comment by aaronbbrown** [comment]
@electrical I changed the error message.  Might want to run the automated tests against it again, just because I made a slight syntax change as well to avoid a 100+ char long line.

**Comment by kevpl** [comment]
@puppetlabs-jenkins retest this please

**Comment by aaronbbrown** [comment]
What was the failure?

**Comment by aaronbbrown** [comment]
@electrical ping. What failed? (I don't have access to Jenkins)

**Comment by electrical** [comment]
@aaronbbrown I'm afraid i don't have access either since i don't work at puppetlabs :-)
@kevpl could you have a look on Monday? Thanks!

**Comment by tvpartytonight** [comment]
@puppetlabs-jenkins retest this please

**Comment by aaronbbrown** [comment]
> @puppetlabs-jenkins retest this please

doesn't seem like he heard you :D

**Comment by tvpartytonight** [comment]
@puppetlabs-jenkins test this please

**Comment by tvpartytonight** [comment]
Apologies @aaronbbrown , I had changed my github username and neglected to updated our jenkins infrastructure. The next kickoff should work.

**Comment by tvpartytonight** [comment]
@puppetlabs-jenkins test this please

**Comment by aaronbbrown** [comment]
Is this good to go?

**Comment by tvpartytonight** [comment]
LGTM @aaronbbrown ; thanks for the PR!

**Pull request merged**